core/src/main/java/jenkins/util/JavaVMArguments.java

```diff
@@ -20,9 +21,16 @@ public class JavaVMArguments {
      * Gets the process argument list of the current process.
      */
     public static List<String> current() {
-        // Native approach
-        if (Functions.isGlibcSupported()) {
-            int pid = GNUCLibrary.LIBC.getpid();
+        ProcessHandle.Info info = ProcessHandle.current().info();
+        if (info.command().isPresent() && info.arguments().isPresent()) {
+            // Java 9+ approach
+            List<String> args = new ArrayList<>();
+            args.add(info.command().get());
+            Stream.of(info.arguments().get()).forEach(args::add);
+            return args;
+        } else if (Functions.isGlibcSupported()) {
```

Comment:
is there a reason the native approach has been kept? both here and in the restarter?

Reply:
I should hope there is a good reason. I would not want to do anything based on sloppy reasoning. While implementing `ProcessHandle.pid()` merely requires a call to `getpid(2)`, there is no standard C or C++ API that could be used to implement `ProcessHandle.info()`: it requires OS-specific code, some of which can be fairly complicated even in C/C++ (and even more so in Java, as can be seen from our `ProcessTree` class). The documentation for [`ProcessHandle.Info`](https://docs.oracle.com/javase/9/docs/api/java/lang/ProcessHandle.Info.html) returns optional types for everything and is very vague about what is supported:

> The attributes of a process vary by operating system and are not available in all implementations. Information about processes is limited by the operating system privileges of the process making the request. The return types are `Optional<T>` allowing explicit tests and actions if the value is available.

While in practice this API has returned valid results during all of my testing on Linux, I would not be surprised if the OpenJDK developers declined to implement this API for some obscure platforms like AIX or Solaris/illumos: platforms which are supported by our `ProcessTree` API, have active users who have complained when we tried to remove support for those platforms, and whose users would be negatively affected by any potential regression caused by the removal of our (working!) `ProcessTree` implementation without a suitable replacement from OpenJDK. The basic point is that the API explicitly states that it cannot be relied on, so we do not rely on it.